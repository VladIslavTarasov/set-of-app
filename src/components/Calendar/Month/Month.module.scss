.table {
  flex-grow: 1;
  border: 2px solid var(--black1);
}

.day {
  position: relative;
  border: 1px solid var(--black1);
  text-align: center;

  &:not(:first-child) {
    border-left: 2px solid var(--black1);
  }
}

.dayHead {
  background-color: var(--black1);
}

.hasTask {
  position: relative;

  &::before {
    position: absolute;
    top: 15%;
    right: 25%;
    width: 10px;
    height: 10px;
    background-color: var(--green1);
    border-radius: var(--radius-circle);
    content: '';
  }
}
