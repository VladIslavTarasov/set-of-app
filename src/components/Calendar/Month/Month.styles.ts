import { createUseStyles } from 'react-jss';

import { Theme } from 'theme/types/theme.types';

export const useStyles = createUseStyles(
  {
    table: () => ({
      flexGrow: 1,
    }),
    row: {
      display: 'flex',
    },
    cell: () => ({
      flex: '1 1 calc(100% / 7)',
      position: 'relative',
      textAlign: 'center',
    }),
    hasEvent: ({ theme }: { theme: Theme }) => ({
      position: 'relative',

      '&::before': {
        content: "''",
        position: 'absolute',
        zIndex: 2,
        left: '50%',
        bottom: 10,
        width: 35,
        height: 5,
        boxSizing: 'border-box',
        transform: 'translateX(-50%)',
        backgroundColor: theme.palette.common.gray,
        borderRadius: theme.borderRadius.xs,
      },
    }),
  },
  { name: 'Month' }
);
