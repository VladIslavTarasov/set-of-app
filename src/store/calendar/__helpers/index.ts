export { updateCalendar } from './updateCalendar';
export { getCalendarData } from './getCalendarData';
