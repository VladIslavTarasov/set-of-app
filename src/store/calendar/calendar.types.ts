export type CalendarDay = {
  readonly day: number;
  readonly date: string;
  readonly isActiveMonth: boolean;
  readonly action?: 'next' | 'prev';
};

export type State = {
  readonly currentMonth: string[];
  readonly currentDay: string;
  readonly calendar: CalendarDay[][];
};

export const NEXT_MONTH = 'CALENDAR/NEXT_MONTH';

export type NextMonthAction = {
  type: typeof NEXT_MONTH;
};

export type NextMonthActionCreator = () => NextMonthAction;

export const PREV_MONTH = 'CALENDAR/PREV_MONTH';

export type PrevMonthAction = {
  type: typeof PREV_MONTH;
};

export type PrevMonthActionCreator = () => PrevMonthAction;

export const CHANGE_DATE = 'CALENDAR/CHANGE_DATE';

export type ChangeDateAction = {
  type: typeof CHANGE_DATE;
  payload: string;
};

export type ChangeDateActionCreator = (year: string) => ChangeDateAction;

export type Actions = NextMonthAction | PrevMonthAction | ChangeDateAction;
