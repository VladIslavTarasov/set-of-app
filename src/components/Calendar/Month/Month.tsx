import React, { memo, useState, useCallback } from 'react';

import cn from 'classnames';
import moment from 'moment';

import DateButton from 'components/Calendar/Day';
import { CalendarDay } from 'store/calendar/calendar.types';

import style from './Month.module.scss';

interface BodyProps {
  onChangeDate: (date: string) => void;
  dates: string[] | null;
  calendar: CalendarDay[][];
  currentDate: string;
  loading?: boolean;
}

const weekDays = moment.weekdaysShort();

const Table: React.FC<BodyProps> = ({ dates, calendar, currentDate, loading, onChangeDate }) => {
  const [choosenDate, setChoosenDate] = useState<string>(currentDate);

  const handleChangeDate = useCallback(
    (date: string) => {
      onChangeDate(date);

      setChoosenDate(date);
    },
    [onChangeDate]
  );

  return (
    <table className={style.table}>
      <thead>
        <tr>
          {weekDays.map(day => (
            <th key={day} className={style.dayHead}>
              {day}
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {calendar.map((weeks, i) => (
          // eslint-disable-next-line
          <tr key={i} className={style.tr}>
            {weeks.map(day => (
              <td
                key={day.date}
                className={cn(style.day, {
                  [style.hasTask]: !loading && dates?.includes(day.date),
                })}
              >
                <DateButton
                  loading={loading}
                  onChangeDate={handleChangeDate}
                  choosenDate={choosenDate}
                  {...day}
                />
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default memo(Table);
