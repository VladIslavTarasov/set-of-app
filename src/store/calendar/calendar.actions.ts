import * as calendarTypes from './calendar.types';

export const toNextMonth: calendarTypes.NextMonthActionCreator = () => ({
  type: calendarTypes.NEXT_MONTH,
});

export const toPrevMonth: calendarTypes.PrevMonthActionCreator = () => ({
  type: calendarTypes.PREV_MONTH,
});

export const changeDate: calendarTypes.ChangeDateActionCreator = (year: string) => ({
  type: calendarTypes.CHANGE_DATE,
  payload: year,
});
