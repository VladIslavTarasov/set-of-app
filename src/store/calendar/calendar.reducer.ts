import moment from 'moment';

import { getCalendarData, updateCalendar } from './__helpers';
import * as calendarTypes from './calendar.types';

const localMoment: moment.Moment = moment();

export const initial: calendarTypes.State = {
  currentMonth: [localMoment.format('YYYY MMMM'), localMoment.format('YYYY-MM')],
  currentDay: localMoment.format('YYYY-MM-D'),
  calendar: getCalendarData(),
};

const calendarReducer = (
  state: calendarTypes.State,
  action: calendarTypes.Actions
): calendarTypes.State => {
  switch (action.type) {
    case calendarTypes.NEXT_MONTH:
      return { ...state, ...updateCalendar(state.currentMonth[0], 'next') };
    case calendarTypes.PREV_MONTH:
      return { ...state, ...updateCalendar(state.currentMonth[0], 'prev') };
    case calendarTypes.CHANGE_DATE:
      return {
        ...state,
        ...updateCalendar(action.payload, 'byDate'),
      };

    default:
      return state;
  }
};

export default calendarReducer;
