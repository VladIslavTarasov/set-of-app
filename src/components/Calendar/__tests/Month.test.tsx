import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import calendar from 'store/calendar/__tests/stubs';

import Month from '../Month';

describe('components/Calendar', () => {
  it('/Month', () => {
    const monthProps = {
      // eslint-disable-next-line
      onChangeDate: (date: string) => {},
      calendar: calendar['2020-05'],
      currentDate: '2020-05-5',
      dates: ['2020-05-5'],
    };

    const { container, rerender, getByTitle } = render(<Month {...monthProps} />);
    expect(container.querySelectorAll('.hasTask').length).toBe(monthProps.dates.length);
    expect(container).toMatchSnapshot();

    rerender(<Month {...monthProps} dates={['2020-05-6', '2020-05-8', '2020-05-11']} />);
    expect(container.querySelectorAll('.hasTask').length).toBe(3);
    expect(container.querySelector('.choosen')?.textContent).toBe('5');

    fireEvent.click(getByTitle('2020-05-14'));
    expect(container.querySelector('.choosen')?.textContent).toBe('14');
    expect(container).toMatchSnapshot();
  });
});
