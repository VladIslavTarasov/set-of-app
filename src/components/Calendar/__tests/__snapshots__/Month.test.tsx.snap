// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Calendar /Month 1`] = `
<div>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          class="dayHead"
        >
          Sun
        </th>
        <th
          class="dayHead"
        >
          Mon
        </th>
        <th
          class="dayHead"
        >
          Tue
        </th>
        <th
          class="dayHead"
        >
          Wed
        </th>
        <th
          class="dayHead"
        >
          Thu
        </th>
        <th
          class="dayHead"
        >
          Fri
        </th>
        <th
          class="dayHead"
        >
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-26"
            type="button"
          >
            26
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-27"
            type="button"
          >
            27
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-28"
            type="button"
          >
            28
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-29"
            type="button"
          >
            29
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-30"
            type="button"
          >
            30
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-1"
            type="button"
          >
            1
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-2"
            type="button"
          >
            2
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-3"
            type="button"
          >
            3
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-4"
            type="button"
          >
            4
          </button>
        </td>
        <td
          class="day hasTask"
        >
          <button
            class="button activeMonth choosen"
            title="2020-05-5"
            type="button"
          >
            5
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-6"
            type="button"
          >
            6
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-7"
            type="button"
          >
            7
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-8"
            type="button"
          >
            8
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-9"
            type="button"
          >
            9
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-10"
            type="button"
          >
            10
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-11"
            type="button"
          >
            11
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-12"
            type="button"
          >
            12
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-13"
            type="button"
          >
            13
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-14"
            type="button"
          >
            14
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-15"
            type="button"
          >
            15
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-16"
            type="button"
          >
            16
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-17"
            type="button"
          >
            17
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-18"
            type="button"
          >
            18
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-19"
            type="button"
          >
            19
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-20"
            type="button"
          >
            20
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-21"
            type="button"
          >
            21
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-22"
            type="button"
          >
            22
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-23"
            type="button"
          >
            23
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-24"
            type="button"
          >
            24
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-25"
            type="button"
          >
            25
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-26"
            type="button"
          >
            26
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-27"
            type="button"
          >
            27
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-28"
            type="button"
          >
            28
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-29"
            type="button"
          >
            29
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-30"
            type="button"
          >
            30
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-31"
            type="button"
          >
            31
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-1"
            type="button"
          >
            1
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-2"
            type="button"
          >
            2
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-3"
            type="button"
          >
            3
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-4"
            type="button"
          >
            4
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-5"
            type="button"
          >
            5
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-6"
            type="button"
          >
            6
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`components/Calendar /Month 2`] = `
<div>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          class="dayHead"
        >
          Sun
        </th>
        <th
          class="dayHead"
        >
          Mon
        </th>
        <th
          class="dayHead"
        >
          Tue
        </th>
        <th
          class="dayHead"
        >
          Wed
        </th>
        <th
          class="dayHead"
        >
          Thu
        </th>
        <th
          class="dayHead"
        >
          Fri
        </th>
        <th
          class="dayHead"
        >
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-26"
            type="button"
          >
            26
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-27"
            type="button"
          >
            27
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-28"
            type="button"
          >
            28
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-29"
            type="button"
          >
            29
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-04-30"
            type="button"
          >
            30
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-1"
            type="button"
          >
            1
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-2"
            type="button"
          >
            2
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-3"
            type="button"
          >
            3
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-4"
            type="button"
          >
            4
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-5"
            type="button"
          >
            5
          </button>
        </td>
        <td
          class="day hasTask"
        >
          <button
            class="button activeMonth"
            title="2020-05-6"
            type="button"
          >
            6
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-7"
            type="button"
          >
            7
          </button>
        </td>
        <td
          class="day hasTask"
        >
          <button
            class="button activeMonth"
            title="2020-05-8"
            type="button"
          >
            8
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-9"
            type="button"
          >
            9
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-10"
            type="button"
          >
            10
          </button>
        </td>
        <td
          class="day hasTask"
        >
          <button
            class="button activeMonth"
            title="2020-05-11"
            type="button"
          >
            11
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-12"
            type="button"
          >
            12
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-13"
            type="button"
          >
            13
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth choosen"
            title="2020-05-14"
            type="button"
          >
            14
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-15"
            type="button"
          >
            15
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-16"
            type="button"
          >
            16
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-17"
            type="button"
          >
            17
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-18"
            type="button"
          >
            18
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-19"
            type="button"
          >
            19
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-20"
            type="button"
          >
            20
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-21"
            type="button"
          >
            21
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-22"
            type="button"
          >
            22
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-23"
            type="button"
          >
            23
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-24"
            type="button"
          >
            24
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-25"
            type="button"
          >
            25
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-26"
            type="button"
          >
            26
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-27"
            type="button"
          >
            27
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-28"
            type="button"
          >
            28
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-29"
            type="button"
          >
            29
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-30"
            type="button"
          >
            30
          </button>
        </td>
      </tr>
      <tr
        class="tr"
      >
        <td
          class="day"
        >
          <button
            class="button activeMonth"
            title="2020-05-31"
            type="button"
          >
            31
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-1"
            type="button"
          >
            1
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-2"
            type="button"
          >
            2
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-3"
            type="button"
          >
            3
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-4"
            type="button"
          >
            4
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-5"
            type="button"
          >
            5
          </button>
        </td>
        <td
          class="day"
        >
          <button
            class="button"
            title="2020-06-6"
            type="button"
          >
            6
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
